lib/ClangImporter/ImportDecl.cpp

```diff
@@ -2486,14 +2486,114 @@ namespace {
         ctors.push_back(valueCtor);
       }
 
-      // Do not allow Swift to construct foreign reference types (at least, not
-      // yet).
       if (isa<StructDecl>(result)) {
+        // For c++ value types, import all the constructors
         for (auto ctor : ctors) {
           // Add ctors directly as they cannot always be looked up from the
           // clang decl (some are synthesized by Swift).
           result->addMember(ctor);
         }
+      } else if (isa<ClassDecl>(result)) {
+        // For C++ foreign reference types, generate a static factory method to
+        // construct the object and import it as a Swift initializer.
+
+        // TODO: better/safer way to do this casting
+        clang::CXXRecordDecl *cxxRecordDecl =
+            (clang::CXXRecordDecl *)cast<clang::CXXRecordDecl>(decl);
+
+        ASTContext &ctx = result->getASTContext();
+        clang::ASTContext &clangCtx = cxxRecordDecl->getASTContext();
+
+        // Creating a name for the new synthesized function
+        clang::IdentifierTable &idents = clangCtx.Idents;
+        clang::IdentifierInfo *generatedFuncName =
+            &idents.get(("returns" + cxxRecordDecl->getNameAsString()).c_str());
+
+        // Creating the return type for the synthesized function to be the
+        // pointe to cxxRecordDecl
+        clang::QualType cxxRecordTy = clangCtx.getRecordType(cxxRecordDecl);
+        clang::QualType cxxRecordPtrTy = clangCtx.getPointerType(cxxRecordTy);
+        // TODO/DOUBT: How to add `_Nonnull` as a qualifier on cxxRecordPtrTy
+        // return type ??
+
+        // Creating the type for the synthesized function
+        clang::FunctionProtoType::ExtProtoInfo EPI;
+        clang::QualType generatedFuncTy =
+            clangCtx.getFunctionType(cxxRecordPtrTy, {}, EPI);
+
+        clang::CXXMethodDecl *generatedCxxMethodDecl =
+            clang::CXXMethodDecl::Create(
+                clangCtx, cxxRecordDecl, clang::SourceLocation(),
+                clang::DeclarationNameInfo(generatedFuncName,
+                                           clang::SourceLocation()),
+                generatedFuncTy, nullptr, clang::SC_Static, false, true,
```

Comment:
For readability, we usually add comments to the bool parameters to see what they stand for, e.g.: `/*UsesFPIntrin=*/false, /*isInline=*/true,`.

Reply:
Let's pass in some `TypeSourceInfo` as well here. E.g., via `clangCtx.getTrivialTypeSourceInfo(..)`.

Reply:
Github collapses comments on outdated code, but just anted to make sure you are still tracking those because there are many unresolved comments, like this one.